src/sign.c

```diff
@@ -315,6 +326,163 @@ int wally_ec_sig_to_public_key(const unsigned char *bytes, size_t bytes_len,
     return ok ? WALLY_OK : WALLY_EINVAL;
 }
 
+#define IS_SCALAR_VALID(s, s_len) (s && s_len == EC_SCALAR_LEN)
+
+int wally_ec_scalar_verify(const unsigned char *scalar, size_t scalar_len)
+{
+    if (!IS_SCALAR_VALID(scalar, scalar_len))
+        return WALLY_EINVAL;
+    return mem_is_zero(scalar, scalar_len) || seckey_verify(scalar) ? WALLY_OK : WALLY_EINVAL;
+}
+
+static bool check_scalar_op_args(const unsigned char *scalar, size_t scalar_len,
+                                 const unsigned char *operand, size_t operand_len,
+                                 unsigned char *bytes_out, size_t len)
+{
+    if (bytes_out && len)
+        wally_clear(bytes_out, len);
+
+    return IS_SCALAR_VALID(scalar, scalar_len) &&
+           IS_SCALAR_VALID(operand, operand_len) &&
+           IS_SCALAR_VALID(bytes_out, len);
+}
+
+int wally_ec_scalar_add(const unsigned char *scalar, size_t scalar_len,
+                        const unsigned char *operand, size_t operand_len,
+                        unsigned char *bytes_out, size_t len)
+{
+    unsigned char tmp[EC_SCALAR_LEN];
+
+    if (!check_scalar_op_args(scalar, scalar_len, operand, operand_len, bytes_out, len))
+        return WALLY_EINVAL;
+
+    if (mem_is_zero(operand, len)) {
+        /* X + 0 = X */
+        if (!mem_is_zero(scalar, scalar_len) && !seckey_verify(scalar))
+            return WALLY_ERROR; /* Outside the group order */
+        memcpy(bytes_out, scalar, len);
+        return WALLY_OK;
+    }
+
+    if (mem_is_zero(scalar, len)) {
+        /* 0 + X = X */
+        if (!seckey_verify(operand))
+            return WALLY_ERROR; /* Outside the group order */
+        memcpy(bytes_out, operand, len);
+        return WALLY_OK;
+    }
+
+    /* Check for addition of the scalars inverse */
+    memcpy(tmp, operand, len);
+    if (!seckey_negate(tmp))
+        return WALLY_ERROR; /* Outside the group order */
+
+    if (!memcmp(scalar, tmp, len)) {
+        /* X + -X = 0 */
+        wally_clear(bytes_out, len);
```

Comment:
Should be cleared already by op_args check, but no biggie (same in subtract below).  
Only mention because add and subtract explicitly clear, but multiply doesn't (and has a comment saying that it's already cleared).